physics/star-formation/include/star_data.hpp

```diff
@@ -0,0 +1,104 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*! @file
+ * @brief Unit tests for ParticlesData
+ *
+ * @author Noah Kubli <noah.kubli@uzh.ch>
+ * @author Sebastian Keller <sebastian.f.keller@gmail.com>
+ */
+#pragma once
+
+#include <array>
+#include <optional>
+
+#include "cstone/fields/enumerate.hpp"
+#include "cstone/fields/field_states.hpp"
+#include "cstone/fields/particles_get.hpp"
+#include "cstone/util/reallocate.hpp"
+
+namespace star_formation
+{
+
+template<class T>
+class StarData : public cstone::FieldStates<StarData<T>>
+{
+public:
+    inline static constexpr size_t numFields = 2;
+
+    template<class ValueType>
+    using FieldVector     = std::vector<ValueType, std::allocator<ValueType>>;
+    using RealType        = T;
+    using AcceleratorType = cstone::CpuTag;
+
+    std::array<FieldVector<T>, numFields> fields;
+
+    auto dataTuple() { return dataTuple_helper(std::make_index_sequence<numFields>{}); }
+
+    auto data()
+    {
+        using FieldType =
+            std::variant<FieldVector<float>*, FieldVector<double>*, FieldVector<unsigned>*, FieldVector<uint64_t>*>;
+
+        return std::apply([](auto&... fields) { return std::array<FieldType, sizeof...(fields)>{&fields...}; },
+                          dataTuple());
+    }
+
+    void resize(size_t size)
+    {
+        double growthRate = 1.05;
+        auto   data_      = data();
+
+        for (size_t i = 0; i < data_.size(); ++i)
+        {
+            if (this->isAllocated(i))
+            {
+                std::visit([size, growthRate](auto& arg) { reallocate(*arg, size, growthRate); }, data_[i]);
+            }
+        }
+    }
+
+    //! Star field names
+    inline static constexpr std::array fieldNames{"t_form",
+            "m_form"};
+
+    static_assert(fieldNames.size() == numFields);
+
+private:
+    template<size_t... Is>
+    auto dataTuple_helper(std::index_sequence<Is...>)
+    {
+        return std::tie(fields[Is]...);
+    }
+};
+
+} // namespace star_formation
+
+#endif // SPHEXA_STAR_DATA_HPP
```

Comment:
I don't think it makes sense to add this yet without any calling or testing code. Better to wait until we have the particle types available.